game-core/src/system/camera.rs

```diff
@@ -16,33 +17,22 @@ impl<'s> System<'s> for Movement {
     );
 
     fn run(&mut self, (players, cameras, mut transforms): Self::SystemData) {
-        let mut player_x: f32 = 0.0;
-        let mut player_y: f32 = 0.0;
-
-        // let mut camera_x: f32 = 0.0;
-        // let mut camera_y: f32 = 0.0;
-
-        // let mut camera_safe_x: f32 = 0.0;
-        // let mut camera_safe_y: f32 = 0.0;
+        let mut player_translation = Vector2 {x: 0.0 as f32, y: 0.0 as f32,};
 
         for (_, transform) in (&players, &mut transforms).join() {
-            player_x = transform.translation.x - 256.0 as f32;// * 5.0;
-            player_y = transform.translation.y - 256.0 as f32;// * 5.0;
+            player_translation = transform.translation.truncate() - Vector2 {x: 256.0, y: 256.0,};
         }
 
         for (_, transform) in (&cameras, &mut transforms).join() {
-            let camera_x = transform.translation.x as f32;// * 5.0;
-            let camera_y = transform.translation.y as f32;// * 5.0;
-            let camera_safe_x = transform.scale.x/4 as f32;
-            let camera_safe_y = transform.scale.y/4 as f32;
-
-            if (player_x - camera_x).abs() > camera_safe_x || (player_y - camera_y).abs() > camera_safe_y {
-                println!("player escaped camera safe zone");
-                transform.translation.x = player_x;
-                transform.translation.y = player_y;
-            }   
-
+            let camera_translation = transform.translation.truncate();
+            let player_direction = player_translation - camera_translation;
+            let camera_safe_edge = player_direction.normalize_to((transform.scale.truncate() / 4 as f32).magnitude());
```

Comment:
just use `4.0` rather than `4 as f32`

Reply:
I don't see why you're calculating a normalized vector based on the scale magnitude, which you're then taking the magnitude square of, it seems like you should be calculating the magnitude outside of the if, using that in the boolean, and then normalizing the vector in the if block below if the camera needs to shift.